design-proposal: Support isolateEmulatorThread with SMT enabled Nodes (#247)
kubevirt/community: design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

+2
odd(1,3,4,5,7,9,...)
+1 (as usual)

Notes:
1. In both scenarios the guest will get the same amount of CPUs it requested.
2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.

Pros:
* Cleaner user experience. No prior knowledge of CPUs added behind the scene needed. 
* Matches the current crio convention, like `cpu-load-balancing.crio.io: disable`
* Reduces the “wasted” CPUs to minimum per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs. 
* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager properly prevent the situation from happening.

Cons:
* We’re adding yet another semi alpha knob to an already complicated set of knobs.

# Fix Alternatives
## Alternative #1 Adding `full-pcpus-only` option as a global configuration
As opposed to the suggested solution where the user needs to add this annotation to every VMI, it is possible to make this behavior global. The assumption is that clusters run homogeneous Nodes, so this will eliminate the need to add this annotation every time.

Pros:

> * On homogeneous Nodes will make the configuration of the VMI less complex.

Reviewer: ```suggestion
* VM configuration is less complex
```

Author: thanks.

Cons:

> * Although some D/S distributions like OCP require that the cluster is homogeneous - it is not a common requirement accepted by the community.

Reviewer: ```suggestion
* The suggested annotation is redundant on nodes with hyperthreading disabled. A cluster-wide configuration may be wasteful for heterogeneous clusters, where node-level granularity is needed.
```

Reviewer: It is important to say that this solution works on heterogeneous clusters. It is just wastes an extra vpcu even on nodes that do not require this waste.

Author: got it. I agree.

In order to avoid noisy CPU neighbors, such VMs need to use the [dedicatedCPUPlacement and IsolateEmulatorThread](https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/#requesting-dedicated-cpu-for-qemu-emulator) features, along with other custom cluster configuration.
One of these cluster changes is to set the Kubelet’s CPUManager Policy to static - [full-pcpus-only](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/). This forces kubelet to only accept Pods that have CPUs with full physical cores.

However, on nodes with [hyperthreading](https://en.wikipedia.org/wiki/Hyper-threading) (SMT) enabled, this causes an error, described on [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103).
This design proposal is meant to fix this issue.

Considering that hyperthreading (SMT) is enabled by default in many [Intel](https://www.intel.com/content/www/us/en/architecture-and-technology/hyper-threading/hyper-threading-technology.html) processors - it means that DPDK and Realtime users may be impacted.

## Motivation

> Running High performance VMs with an even number of CPUs with the conditions below.

Reviewer 2: That includes guests with SMT enabled (spec.domain.cpu.threads: 2).

Author: yes.

Reviewer 2: > That's not needed. You can turn off SMT on the guest `spec.domain.cpu.threads: 1`, and still get the SMT issue on the pod.

I gave an example of a specific scenario that is not working.
Any even combination of sockets, cores and threads will cause this problem.

Author: I don't follow. Is there something we need to add here to make things clearer?

Author: After talking offline - resolving issue

          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
kubelet will emit following event:
```
SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
```
The following picture helps show how this is a problem
<img src="smt-slignment-error-diagram.png">

## Goals
Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.

## Non  Goals
Fix the actual scheduling of the virt-launcher to a node that will reject it.

## Definition of Users
This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories
A user should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)

# Bug fix design proposal
The bug fix proposal introduces an alpha annotation `kubevirt.io/full-pcpus-only: true`:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    metadata:
      annotations:

>         kubevirt.io/full-pcpus-only: true

Reviewer 2: Please consider having an annotation without a value.

Author: It fits the convention of `isolateEmulatorThread: true` (though this one is an API)
So I can go either way. 
Any objections to setting the annotation to `kubevirt.io/full-pcpus-only:`?

Reviewer 3: Perhaps, it would be better to match the CPU manager's policy wording and change this to something like, `kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only`
This way it will be clear where this is coming from and may accommodate other policies there might (in the future - hopefully we won't get to this)

Author: DONE

    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```

When the VM is configured with `isolateEmulatorThread: true` and the above annotation, the virt-controller will set the virt-launcher pod with an even amount of CPUs, meaning:

| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) |
|-----------------------------------|------------------|-----------------------|
| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         |

Notes:
1. In both scenarios the guest will get the same amount of CPUs it requested.

> 2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.

Reviewer 2: If a dedicated CPU is already present on the virt-launcher pod, but not used - it is a waste, as no other code could run on it.

Author: true. But notice that virt-hanlder will have to check
- at the VM level: the existence of this annotations
- at the host level:  if this CPU has a sibling, and what it is.
It's a more complex, and possible has more implications than we think. so I'm leaving it fir debate.

@vladikr @iholder101 Can you share your thoughts ?

Reviewer 3: I agree, in general...
The emulator thread is being pinned to a CPUSet [1]
And this CPUSet is being used to build the housekeeping cgroup [2].

I think it shouldn't be too hard to update [1] to place the 2 over-allocated CPUs in this set.

[1] https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go#L364
[2] https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L2828

Pros:
* This option does not introduce any alpha knobs.

Cons:
* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.
* KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.
* VMs that were defined like this will have to change when the long term solution will arrive (backwards-compatibility issue)

## Scalability
The solution proposed is “wasting” one unused CPU for the purpose of maintaining SMT alignment. I believe it is a minor waste considering the alternatives, such as disabling SMT on the node. As I consider the solution adequately scalable.

## Update/Rollback Compatibility
Since the annotation is not an API change, it is backportable.
When a better solution is possible, Kubevirt can simply replace the annotation support with a better approach, thus not making the change relatively seamless for the client. 

## Functional Testing Approach

> Will add unit tests to `renderresources_test.go` to check extra housekeeping logic, and in `templates.go` to check the virt-launcher pod spec output.

Reviewer 2: Does an e2e test is also needed?

Author: Good question. 
I think we will encounter a problem with this, since there is no support for SMT on kubevirtci, nor is there a convenient way to set the cpu-manager Policy there. 
Simply put - we can't replicate the SMT alignment error. Hence, I would only add unit tests.
@vladikr what do you think?

Reviewer 3: The main issue is not the SMT support on the nodes, but rather the k8s CPU manager beta policy configuration. I'm not sure how we can have such a configuration in kubevirt CI without creating a dedicated lane.
@xpivarc what do you think?

Reviewer 4: @vladikr  Don't we need both "hardware" and Kubelet configuration?

Reviewer 3: @xpivarc the smt can be adjusted. I imagine that we can do the same trick we did before. We can set one of the nodes to enable smt.
Theoretically, we could change the kubelet configuration, but I fear that it will be tricky since the dynamic configuration has been deprecated.

Reviewer 4: We could but I am not sure about the ROI. Maybe unit tests are enough here?

Reviewer: +1 for unit tests only unless e2e is absolutely required.

Author: IMO all the changes done on the first phase PR are covered by unit tests, making the e2e non-required. 
Regarding the second phase (cgroups PR WIP) - that part is less covered by unit tests so I'll try to refactor it so that we don't need any e2e tests.

Reviewer 3: I don't mind if we will have unit tests only with this feature.
In the general case, unit tests are not enough. Most of the code is never exercised together so there is no way to verify the whole flow.

This looks more as an esoteric use case that relies on a feature-gated option of the CPU manager that is not likely to be widely used upstream and has no plans to get scheduler support.

Reviewer 4: Ok, I think we have an agreement for this one.

| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) |
|-----------------------------------|------------------|-----------------------|
| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         |

Notes:
1. In both scenarios the guest will get the same amount of CPUs it requested.
2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.

Pros:
* Cleaner user experience. No prior knowledge of CPUs added behind the scene is needed. 
* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs. 
* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.

Cons:
* We’re adding yet another semi alpha knob to an already complicated set of knobs.

# Fix Alternatives

> ## Alternative #1 Adding `full-pcpus-only` option as a global configuration

Reviewer 5: @RamLavi I am not sure whether we can consider it as alternative, because it depends on the main proposal. It can rather be an option in the implementation details, to enhance the kubevirt CR with the option to annotate every VM with the suggested annotation.

Author: >@RamLavi I am not sure whether we can consider it as alternative, because it depends on the main proposal. It can rather be an option in the implementation details, to enhance the kubevirt CR with the option to annotate every VM with the suggested annotation.

I don't mean that this global option will annotate every VM once it is enabled. 
I mean that this will be a global behavior, in a similar way that `defaultRuntimeClass` sets the runtimeClass without changing the VM manifest (metadata or spec API) at all.
It is in that sense - that it is a different alternative.

As opposed to the suggested solution where the user needs to add this annotation to every VM, it is possible to make this behavior global. The assumption is that clusters run homogeneous Nodes, so this will eliminate the need to add this annotation every time.

Pros:
* VM configuration is less complex.
* Similar to the way default runtimeclass is configured on Kubevirt.

Cons:

> * The suggested annotation is waisting a CPU when used on nodes with hyperthreading disabled. A cluster-wide configuration may be wasteful for heterogeneous clusters, where node-level granularity is needed.

Reviewer 5: @RamLavi IIUC the main proposal also has this disadvantage. I can annotate the VM and it can still be scheduled on a non-SMT node in case the cluster is heterogeneous

Author: yes but notice the scope - iIn the main proposal the scope is per VM, hence minimal (only when necessary), whereas on this alternative the impact is every VM on the cluster (!).
Considering that a standard Node has a few dozens of CPUs, this change of scope is what makes this Con - a very big and waistfull one (CPU wise).

> ## Alternative #2 Configure both pod and Guest

Reviewer 5: @RamLavi This is also doesn't look like an alternative solution, but a suggestion to how to implement the addition of the core so that it will be even.

Author: I kinda agree with you here. 
If this alternative is chosen then the Action item is no-op in kubevirt's perspective. I just wanted this option to be heard, as it has its advantages.

Reviewer: Thank you for putting it writing, as this is my favorite alternative.

* VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
* VM requests an even number of CPUs.

When enabling `IsolateEmulatorThread` on a VM with even CPU count, Kubevirt is requesting one extra CPU for the virt-launcher pod. This is an internal change that is not exposed to the user.
After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
For example, let’s consider this VM spec:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
kubelet will emit following event:
```
SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
```
The following picture helps show how this is a problem
<img src="smt-slignment-error-diagram.png">

## Goals
Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.

## Non  Goals

> Fix the actual scheduling of the virt-launcher to a node that will reject it.

Author: rephrased, thanks!

## Definition of Users
This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories
* As a cluster admin, I want to set the cluster-wide IsolatedEmulatorThread behavior for Kubevirt by setting a knob on HCO.

> * As a Kubevirt user, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

Author: Fixed to `As a Kubevirt VM owner`.
DONE

* Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
* VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
* VM requests an even number of CPUs.

When enabling `IsolateEmulatorThread` on a VM with even CPU count, Kubevirt is requesting one extra CPU for the virt-launcher pod. This is an internal change that is not exposed to the user.
After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
For example, let’s consider this VM spec:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
kubelet will emit following event:
```
SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
```
The following picture helps show how this is a problem
<img src="smt-slignment-error-diagram.png">

## Goals

> Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.

Reviewer 2: nit: The BZ was migrated to https://issues.redhat.com/browse/CNV-31584.

* As a cluster admin, I want to enable VMs with an even number of CPUs, `dedicatedCpuPlacement` and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
* As a Kubevirt VM owner, I should be able to run VM with isolateEmulatorThread and dedicatedCpuPlacement enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)

# Bug fix design proposal
The bug fix proposal introduces an alpha annotation to the kubevirt CR: `alpha.kubevirt.io/EmulatorThreadCompleteToEvenParity`:
```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  annotations:
    alpha.kubevirt.io/EmulatorThreadCompleteToEvenParity:
spec:
  configuration:
    developerConfiguration:
      featureGates:
      - AlignCPUs
```
This annotation will be respected only when the `AlignCPUs` featureGate is enabled.

## Virt-API:

### VMI mutator webhook
When the following conditions are met:
1. The feature gate is enabled on the Kubevirt CR
2. Kubevirt CR has the annotation.

> 3. VMI is configured with `isolateEmulatorThread: true` and `dedicatedCpuPlacement: true`

Reviewer 2: The actual [implementation](https://github.com/kubevirt/kubevirt/commit/99c4ec652ff523572563aa104a6cc1142011e69e#diff-a1c4b8c0c3279e3289441a28c20645319f7f873d099fd755cb1de6302fd391c4R100) only took `isolateEmulatorThread` into account, since there is a validating web hook that runs after the mutating webhook to verity both exist.

Usage of the annotation will be behind the feature-gate.

Pros:
* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs.

Cons:
* We’re adding yet another semi alpha knob to an already complicated set of knobs that each VM owner needs to be aware of.

## Alternative #2 Configure both pod and Guest
It is possible to use both the current Guest API and the old yet existing pod API, like so:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        resources:
          requests:
            cpu: '10'
          limits:
            cpu: '10'
        cpu:
          cores: 4
          sockets: 1
          threads: 2
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
```
In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.

Pros:
* This option does not introduce any alpha knobs.

Cons:

> * This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.

Reviewer: ```suggestion
* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse.
```

I think that there's a lot of sense in having less validations.

Author: DONE

After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
For example, let’s consider this VM spec:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
kubelet will emit following event:
```
SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
```
The following picture helps show how this is a problem
<img src="smt-slignment-error-diagram.png">

## Goals
Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.

## Non  Goals
Fix the actual scheduling of the virt-launcher to a node that will reject it.

## Definition of Users
This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories

> * As a cluster admin, I want to set the cluster-wide IsolatedEmulatorThread behavior for Kubevirt by setting a knob on HCO.

Author: I see what you mean now. 
DONE

> * As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.

Author: DONE

* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)

# Bug fix design proposal

> The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:

Author: DONE

```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  annotations:
    kubevirt.io/EmulatorThreadCompleteToEvenParity:
spec:
  configuration:
    developerConfiguration:
      featureGates:

>       - EmulatorThreadCompleteToEvenParity

Author: DONE

```
This annotation will be respected only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.

## Virt-API:

### VMI mutator webhook
When the following conditions are met:
1. The feature gate is enabled on the Kubevirt CR
2. Kubevirt CR has the annotation.
3. VMI is configured with `isolateEmulatorThread: true`

then the VMI mutator webhook will copy the annotation from the kubevirt CR to the VMI annotations. This is done in order to pass the information between the virt-controller, virt-handler and virt-launcher, without changing the API between the virt-handler and the virt-launcher.

### VMI create-admitter validation webhook

> Should allow the VMI to have the new annotation only if the feature gate is enabled on the Kubevirt CR

Reviewer 4: me & @vladikr agreed that for now, it is ok if the VMI is going to be admitted to the system. We will discuss guidance on future features and when this feature gets a formal API.

Author: So.. entirely remove this check from the create-admitter validation webhook, right?

## Virt-controller:
When the following conditions are met:

> 1. VMI is configured with `isolateEmulatorThread: true`

Reviewer 4: \+ `dedicatedCpuPlacement`

Author: DONE

2. VMI has the annotation.

then virt-controller will set the virt-launcher pod with an even amount of CPUs, meaning:

| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) | Notes                                                                       |
|-----------------------------------|------------------|-----------------------|-----------------------------------------------------------------------------|

> | Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of CPUs it requested |

Reviewer 4: ```suggestion
| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of vCPUs it requested |
```

Author: DONE

## Virt-launcher:
When the following conditions are met:

> 1. VMI is configured with `isolateEmulatorThread: true`

Reviewer 4: Same as above.

Author: DONE

2. VMI has the annotation.

> then virt-launcher pod will then assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin.

Reviewer 4: ```suggestion
virt-launcher pod will then assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin.
```

Author: DONE

Pros:
* Cleaner user experience for the VM owner. No prior knowledge of CPUs added behind the scene is needed.
* VM configuration is less complex.
* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.

Cons:
* Setting an annotation in kubevirt CR is unconventional. 
* A cluster-wide configuration such as this may be CPU-wasteful for heterogeneous clusters, where node-level granularity is needed.

# Fix Alternatives
## Alternative #1 Adding `EmulatorThreadEvenParity` annotation per VMI
Another way to do it is by adding the annotation on the VMI's annotations:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    metadata:
      annotations:
        kubevirt.io/EmulatorThreadEvenParity:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
Setting this will set the same behavior as on the suggested solution on virt-launcher and virt-handler.

> As opposed to the suggested solution where the user needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This removes the assumption that the Nodes are homogeneous.

Reviewer: ```suggestion
As opposed to the suggested solution where the cluster admin needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This would make it possible for owners of VM with even-numbered CPU, to run them on non-homogeneous clusters without wasting one CPU per VM.
```

There is no *requirement* to have a homogeneous cluster in either alternative. The only thing that different is who we are optimizing for.

Author: ACK. 
DONE.

The annotation will still be accompanied by the feature-gate configuration.

Pros:
* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs.

> * As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.

Reviewer: ```suggestion
```
The chosen suggestion also has this quality.

Author: DONE

Cons:

> * We’re adding yet another semi alpha knob to an already complicated set of knobs.

Reviewer: ```suggestion
* We’re adding yet another semi alpha knob to an already complicated set of knobs that each VM owner needs to be aware of.
```

Author: DONE

## Alternative #2 Configure both pod and Guest
It is possible to use both the current Guest API and the old yet existing pod API, like so:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        resources:
          requests:
            cpu: '10'
          limits:
            cpu: '10'
        cpu:
          cores: 4
          sockets: 1
          threads: 2
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
```
In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.

Pros:
* This option does not introduce any alpha knobs.

Cons:

> * This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.

Reviewer: If we did not have the webhook, we would not have needed this proposal at all. VM owners that want a specific size for their virt-launcher could have done so themselves, just like that can already do for VMs without isolatedEmulatorThreads.

None of us knows how much misuse this webhook prevented, but we do know how much effort it consts us now. IMO this is prime example where over-protection is harmful.

Author: I think the intention of the original feature developers was to "make sure no one is using it in a way that it was not intended", thus opening a way for bugs that come from miss use. 

However, I do see your point here. Perhaps we could all benefit from sharing this example to the team so that we could converse and align with this for future features.

> * KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.

Reviewer: ```suggestion
* Implementation difficulty: KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.
```

Author: DONE

> * VMs that were defined like this will have to change when the long term solution will arrive (backwards-compatibility issue)

Reviewer: ```suggestion
* VMs that were defined like this will have to change when the long term solution will arrive (backwards-compatibility issue)
```
Why? Do you think that VMI.spec.resources.requests.cpu would ever be deprecated?

Author: TBH I don't remember why this was written so perhaps it's better removed.
DONE.

## Non Goals
Set the virt-launcher Pod scheduling to match a node with SMT aligned to its request.

## Definition of Users
This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories
* As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)

# Bug fix design proposal
The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  annotations:
    kubevirt.io/EmulatorThreadCompleteToEvenParity:
spec:
  configuration:
    developerConfiguration:

>       featureGates:

Reviewer 6: Having a feature gate to protect non-API (an annotation) seems redundant and arbitrary. Not having a feature gate would also allow you to remove the webhook checking that a VM annotation is only set if the CR annotation is set. Please consider removing it.

Author: That's a practice commonly practiced on kubevirt (for example [IgnitionEnabled](https://github.com/kubevirt/kubevirt/blob/7e3aa9fff0226fea881fa4efc7906bad3cfcf4be/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L1682)) and was specifically requested by @xpivarc.
@xpivarc - Do you think we can let go of the FG check from the create-admitter validation webhook?

      - EmulatorThreadCompleteToEvenParity
```
This annotation will be respected only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.

## Virt-API:

### VMI mutator webhook
When the following conditions are met:
1. The feature gate is enabled on the Kubevirt CR
2. Kubevirt CR has the annotation.
3. VMI is configured with `isolateEmulatorThread: true`

then the VMI mutator webhook will copy the annotation from the kubevirt CR to the VMI annotations. This is done in order to pass the information between the virt-controller, virt-handler and virt-launcher, without changing the API between the virt-handler and the virt-launcher. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.

### VMI create-admitter validation webhook
Should allow the VMI to have the new annotation only if:

> 1. The feature gate is enabled on the Kubevirt CR

Reviewer 6: We want users to control this feature on the cluster-level through the CR. We should not care if they set it on VM-level using an undocumented annotation. Consider removing this check together with the gate.

Author: I tend to agree.

Reviewer 2: @phoracek please note that with the current design, removing this check will bypass the feature gate on the VMI level - and the behavior change caused by the annotation will take affect.

Reviewer: I'm for *having* a feature gate controlling the new functionality. But we should not reject a VM with an annotation just because a FG is not set. Today if I set `EmulatorThreadCompleteToEvenParity=true` on my VM it is admitted to the system. This should should not change in the future. Admission webhooks should reject VMs with invalid fields (e.g a feature-gated field when the gate is closed). But annotations are lightweight. They are not subject to the same structure scrutiny as actual fields. Typically, any RC owner can create any annotation on their RC without risk of rejection.

Reviewer 4: Annotations with kubevirt subdomain shouldn't be used outside of kubevirt. What would be the behavior if we do not reject the annotated VM when the FG is not set? Would you want an FG check all over the codebase? I don't want it because the result of such a check could change over time and we could quickly get into a weird state. Therefore if a VM has this annotation it should behave as described even if the FG is off in the future.

Reviewer 6: > What would be the behavior if we do not reject the annotated VM when the FG is not set?

Undefined. This would require us to remove the documentation of the per-VM annotation from https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/#requesting-dedicated-cpu-for-qemu-emulator. Then whoever would use the per-VM annotation would be doing it on their own risk.

With this we would have:
* One less feature gate
* One less VM-level knob
* One less webhook check

>  Would you want an FG check all over the codebase?

No. What I suggested is not having the VM-level annotation checked against the feature gate at all.

Reviewer 4: The annotation is not released. The documentation is lying. Nothing is holding us to remove the per-VM documentation if we decide we want to support only cluster-wide knob.

>Undefined.

Doesn't sound as safe and sane as I would like. Even if this is experimental, the behavior or feature overall should be complete in some sense (even in a minimalistic sense) for users to provide at least some relevant feedback.

Again, we have feature gates and we described their meaning well today. I don't see any reason for not using them. The less, less, less argument is not appealing enough for me if in the end, we will introduce the feature gate eventually.

>>   Would you want an FG check all over the codebase?
>No. What I suggested is not having the VM-level annotation checked against the feature gate at all.

So only the presence of the annotation would decide the semantic? That would be fine but again, we have never defined the lifecycle of annotations and we used feature gates for everything else. I don't see a compelling reason to diverge now.

Reviewer 6: I think you got my point, and although I disagree, I won't block this PR anymore if you insist on using webhooks and feature gates here.

Reviewer 4: @phoracek just to be sure, we can do this in the future if we decide on this as a project. But otherwise, we should follow our own rules &/ consistent. We already have a few more examples of how to do one thing than we should have.

Reviewer: > Would you want an FG check all over the codebase?

Yes. I think that it is important. If we decide that a gated feature is going to be dropped, we would like to delete all relevant code. Thus I think that the each dependent behavior should have have a very clear `if (FG==true)` clause.

Much like @phoracek I don't block on this. I just don't like the fact that it imposes non-eventually-consistent behavior.

Reviewer 4: >Yes. I think that it is important. If we decide that a gated feature is going to be dropped, we would like to delete all relevant code. Thus I think that the each dependent behavior should have have a very clear if (FG==true) clause.

I disagree, not only does it complicate the code but it also increases the chance of the system converging to an undefined state. Once a VM with some feature is admitted into the system, the system should be able to handle it. Of course, if we are talking about alpha/beta or gated functionality, we might not guarantee that all cases are covered or for example, upgrades will not break the system.

Reviewer 4: >I don't see what exactly the feature gate is protecting here. It's only the admin that can set the cluster-wide functionality anyway - using the CR annotation.
On a VM level, yes, the user can overallocate another CPU in a very specific case - I don't see an issue here.

What about accidental usage? (copy-pasting from the internet? vendor vs Kubevirt? and so on) I mean if I take these arguments I have a hard time figuring out why we have feature gates. Did we forget about why feature gates exist? For example as an admin, I would like to have control over deployed systems. Or I would like to know when I use something I should have not.

>We should introduce a feature gate only when the is an actual API. Then the feature gate will be checked in the virt-config, not in the virt-launcher.
As long as this functionality is using an experimental annotation I don't see a need for a feature gate.

If you think so, please let`s write this down for future guidance but otherwise, I can find you at least one counter-example.

>I wouldn't reject VMIs as well. The feature gate should control a specific functionality, in the worst case it shouldn't just allocate an additional CPU (that would lead to a cpu manager rejection anyway in the discussed case)

I apologize but I do not follow. The feature gate is controlling a specific functionality or not? I also don't understand the part about the worst case.

Reviewer 3: > > I don't see what exactly the feature gate is protecting here. It's only the admin that can set the cluster-wide functionality anyway - using the CR annotation.
> > On a VM level, yes, the user can overallocate another CPU in a very specific case - I don't see an issue here.
> 
> What about accidental usage? (copy-pasting from the internet? vendor vs Kubevirt? and so on) I mean if I take these arguments I have a hard time figuring out why we have feature gates. Did we forget about why feature gates exist? For example as an admin, I would like to have control over deployed systems. Or I would like to know when I use something I should have not.
> 

In my view, feature gates are generally needed to protect against the use of unstable features / APIs or to enable/disable functionality on a cluster that doesn't have another control.
In this case, we don't have an API, on the other hand, this is not a complete feature either.  This is an extension to an existing logic of the `isolateEmulatorThread`.
By setting this special annotation the admin provides us with consent to enable this logic on a cluster level.

> > We should introduce a feature gate only when the is an actual API. Then the feature gate will be checked in the virt-config, not in the virt-launcher.
> > As long as this functionality is using an experimental annotation I don't see a need for a feature gate.
> 
> If you think so, please let`s write this down for future guidance but otherwise, I can find you at least one counter-example.

I'm sure there are more examples, but in this case, I don't think the feature gate provides any benefits.
From my pov, the feature gate could be incorporated into virt-config function, when we will have a formal API. 

> 
> > I wouldn't reject VMIs as well. The feature gate should control a specific functionality, in the worst case it shouldn't just allocate an additional CPU (that would lead to a cpu manager rejection anyway in the discussed case)
> 
> I apologize but I do not follow. The feature gate is controlling a specific functionality or not? I also don't understand the part about the worst case.

The current annotation extends an existing functionality. If set and other conditions are met, the VMI will allocate an additional CPU, so that the emulator thread will be bound to two cpus. 
If not set, one CPU will be allocated for the emulator thread. 

What I am trying to say is that even if we'll decide to keep the feature gate, it should affect the number of allocated CPUs and not prevent the creation of VMIs. 

I hope it makes sense.

> 2. VMI is configured with `isolateEmulatorThread: true`

Reviewer 6: This can be checked down the stack and fail in the controller or while instantiating the VM. Consider removing this webhook

Author: I agree - it can, but it would require us to patch it. 
The reason the validating webhook was inserted was to only need to check the FG existence there, and make checking it later in the controller unnecessary, as advised by Lubo. I think we should settle it with him on the thread above first.

Author: Also note that rejecting the VMI here if the FG is disabled allows us to not check it on the virt-launcher. Not doing so requires us to pass the FG boolean via the GRPC. Not wanting to alter the GRPC with something the is alpha and might change in the future was also a reason to keep this webhook

Reviewer 6: Is the only alternative checking it on virt-launcher? Can't you check it on virt-api controller level instead of the webhook?

Author: you have to check the annotation twice:
1. On the virt-controller, when setting extra CPUs for the emulator thread.
2. On the virt-launcher, when setting the CPUs for the emulatorThreadPin (will later set on the Housekeeping cgroup).

If we remove the validatin webhook - then we need to also check the FG on these occasions. As I mentioned - the accepted way to do this is the GRPC (see [example](https://github.com/kubevirt/kubevirt/blob/7e3aa9fff0226fea881fa4efc7906bad3cfcf4be/pkg/handler-launcher-com/cmd/v1/cmd.proto#L100)).

Author: IIUC saying "alpha annotation" implies that there is a graduation policy.
@xpivarc we need some clarifications here. Does an annotation require a FG?

Reviewer 4: I would stay consistent because we never described the lifecycle of annotations :) We have a "stable" annotation and yet we do not have a graduation policy. For now, unfortunately, the only option that gates a behavior and is described are feature gates.

## Virt-launcher:
When the following conditions are met:
1. VMI is configured with `isolateEmulatorThread: true`
2. VMI has the annotation.

then virt-launcher pod will then assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin.

## Virt-handler:
After virt-launcher set the emulatorThreadPin, the virt-handler will then set the emulator Threads from the emulatorThreadPin to the housekeeping cgroup.

Pros:
* Cleaner user experience for the VM owner. No prior knowledge of CPUs added behind the scene is needed.
* VM configuration is less complex.
* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.

Cons:
* Setting an annotation in kubevirt CR is unconventional. 
* A cluster-wide configuration such as this may be CPU-wasteful for heterogeneous clusters, where node-level granularity is needed.

# Fix Alternatives
## Alternative #1 Adding `EmulatorThreadCompleteToEvenParity` annotation per VMI

> Another way to do it is by adding the annotation on the VMI's annotations:

Author: DONE

```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    metadata:
      annotations:
        kubevirt.io/EmulatorThreadCompleteToEvenParity:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
This is the same behavior as the selected proposal, without the automatic annotation that is done by the VMI mutator.
As opposed to the suggested solution where the cluster admin needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This would make it possible for owners of VM with even-numbered CPU, to run them on non-homogeneous clusters without wasting one CPU per VM.

> The annotation will still be accompanied by the feature-gate configuration.

Author: DONE

## Functional Testing Approach
Will add unit tests to `renderresources_test.go` to check extra housekeeping logic, and in `templates.go` to check the virt-launcher pod spec output.

# Implementation Phases
## Phase 1: Align CPU to SMT requirements when feature gate is enabled and annotation is set on the VMI.
* Virt-config:
  * Add support for the new feature gate: `EmulatorThreadCompleteToEvenParity`.
* Virt-API:
  * Add support for the new alpha annotation `kubevirt.io/EmulatorThreadCompleteToEvenParity:`.
  * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:
    * the VMI mutator webhook will copy the kubevirt CR annotation to the VMI annotations.
  * in case the VMI has the annotation and either the feature gate is disabled or the isolateEmulatorThread is disabled:
    * the VMI creation will be blocked. 
* Virt-controller:
  * in case isolateEmulatorThread is enabled and the annotation is set on the VMI:
    * Add logic to Align the total number of CPUs to be even.

> ## Phase 2: Adjust housekeeping cgroup to use the extra CPU added by the alignment, if such a one is available.

Reviewer: Why is phase 2 important? Is single CPU typically not more than enough for housekeeping?

Reviewer: Thanks for the pointer. I would add a sentence about the reason in the text of the proposal. My opinion - phase 2 does not worth our trouble.

Author: I understand your view, but note that this part is mostly already merged: 
- https://github.com/kubevirt/kubevirt/pull/10783 merged
- https://github.com/kubevirt/kubevirt/pull/10839 small fix lgtmed, waiting for approval

do you think it is worth reverting?

Reviewer: No. I don't see this as particularly harmful. Just more code for us to maintain, with negligible benefit.

Author: As my suggestion to keep this optional/not-a-must-for-now was rejected, it seems it is not the opinion of our maintainers.

Reviewer: Please add a text here explaining why it was important

Author: At the time @vladikr thought that not incorporating  the second CPU for cgroups is a waist of CPU power.

# Preface
High performance VMs, such as VMs using [DPDK](https://en.wikipedia.org/wiki/Data_Plane_Development_Kit) or [realtime](https://en.wikipedia.org/wiki/Real-time_computing) applications are sensitive to interference from neighboring workloads.
In order to avoid noisy CPU neighbors, such VMs need to use the [dedicatedCPUPlacement and IsolateEmulatorThread](https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/#requesting-dedicated-cpu-for-qemu-emulator) features, along with other custom cluster configuration.

> One of these cluster changes is to set the Kubelet’s CPUManager Policy to static - [full-pcpus-only](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/). This forces kubelet to only accept Pods that have CPUs with full physical cores.

Reviewer 2: ```suggestion
One of these cluster changes is to set the Kubelet’s CPUManager Policy to static - [full-pcpus-only](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/). This forces kubelet to only accept Pods that will occupy full physical cores.
```

Author: DONE

After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
For example, let’s consider this VM spec:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```
kubelet will emit following event:
```
SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
```
The following picture helps show how this is a problem
<img src="smt-slignment-error-diagram.png">

## Goals
Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.

## Non Goals
Set the virt-launcher Pod scheduling to match a node with SMT aligned to its request.

## Definition of Users
This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories

> * As a cluster admin, I want to let `isolatedEmulatorThread` VMs with an even number of CPUs to run on my cluster by setting a cluster-wide knob on HCO.

Reviewer 2: ```suggestion
* As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
```

Author: better..
DONE

* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)

# Bug fix design proposal
The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  annotations:
    kubevirt.io/EmulatorThreadCompleteToEvenParity:
spec:
  configuration:
    developerConfiguration:
      featureGates:
      - EmulatorThreadCompleteToEvenParity
```

> This annotation will be supported only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.

Author: DONE

## Virt-API:

### VMI mutator webhook
When the following conditions are met:
1. The feature gate is enabled on the Kubevirt CR
2. Kubevirt CR has the annotation.
3. VMI is configured with `isolateEmulatorThread: true`

then the VMI mutator webhook will

> 1. Copy the annotation from the kubevirt CR to the VMI annotations. This is done in order for it to be used on the virt-launcher, when setting the housekeeping cgroup. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.

Reviewer 2: No need for numbering here because there is only one step.
> This is done in order for it to be used on the virt-launcher, when setting the housekeeping cgroup.

This is done in order to pass the information between the virt-controller, virt-handler and virt-launcher, without changing the API between the virt-handler and the virt-launcher.

Author: like.
DONE

### VMI create-admitter validation webhook
Should allow the VMI to have the new annotation only if:
1. The feature gate is enabled on the Kubevirt CR
2. VMI is configured with `isolateEmulatorThread: true`

## Virt-controller:
When the following conditions are met:
1. VMI is configured with `isolateEmulatorThread: true`
2. VMI has the annotation.

then virt-controller will:

> 1. Set the virt-launcher pod with an even amount of CPUs, meaning:

Reviewer 2: No need for numbering because there is only one step.

Author: DONE

| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) | Notes                                                                       |
|-----------------------------------|------------------|-----------------------|-----------------------------------------------------------------------------|
| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of CPUs it requested |

## Virt-launcher:
When the following conditions are met:
1. VMI is configured with `isolateEmulatorThread: true`
2. VMI has the annotation.

then virt-launcher pod will then:

> 1. Assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin

Reviewer 2: No need for numbering because there is only one step.

Author: DONE

## Virt-handler:
After virt-launcher set the emulatorThreadPin, the virt-handler will then:

> 1. Set the emulator Threads from the emulatorThreadPin to the housekeeping cgroup.

Reviewer 2: No need for numbering because there is only one step.

Author: DONE

Pros:
* Cleaner user experience for the VM owner. No prior knowledge of CPUs added behind the scene is needed.
* VM configuration is less complex.
* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.

Cons:
* Setting an annotation in kubevirt CR is unconventional. 
* A cluster-wide configuration such as this may be CPU-wasteful for heterogeneous clusters, where node-level granularity is needed.

# Fix Alternatives
## Alternative #1 Adding `EmulatorThreadCompleteToEvenParity` annotation per VMI
Another way to do it is by adding the annotation on the VMI's annotations:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    metadata:
      annotations:
        kubevirt.io/EmulatorThreadCompleteToEvenParity:
    spec:
      domain:
        cpu:
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
          sockets: 1
          cores: 4
          threads: 2
```

> Setting this will set the same behavior as on the suggested solution on virt-launcher and virt-handler.

Author: Changes to `This is the same behavior as the selected proposal, without the automatic annotation that is done by the VMI mutator.`

DONE

Pros:
* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs.

Cons:
* We’re adding yet another semi alpha knob to an already complicated set of knobs that each VM owner needs to be aware of.

## Alternative #2 Configure both pod and Guest
It is possible to use both the current Guest API and the old yet existing pod API, like so:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  running: true
  template:
    spec:
      domain:
        resources:
          requests:
            cpu: '10'
          limits:
            cpu: '10'
        cpu:
          cores: 4
          sockets: 1
          threads: 2
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
```
In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.

Pros:
* This option does not introduce any alpha knobs.

Cons:

> * This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.

Reviewer: In other words, loosening (or even dropping) the validation makes a lot of sense to me. Indeed, it can let people start sub-optimal VMs, but there may be other conditions that would benefit. In no way we get a broken VM or harm to the guest workload, just wasted resources.

## Scalability
The solution proposed is “wasting” one unused CPU for the purpose of maintaining SMT alignment.

## Update/Rollback Compatibility
Since the annotation is not an API change, it is backportable.
When a better solution is possible, Kubevirt can simply replace the annotation support with a better approach, thus not making the change relatively seamless for the client. 

## Functional Testing Approach
Will add unit tests to `renderresources_test.go` to check extra housekeeping logic, and in `templates.go` to check the virt-launcher pod spec output.

# Implementation Phases
## Phase 1: Align CPU to SMT requirements when feature gate is enabled and annotation is set on the VMI.
* Virt-config:
  * Add support for the new feature gate: `EmulatorThreadCompleteToEvenParity`.
* Virt-API:
  * Add support for the new alpha annotation: `kubevirt.io/EmulatorThreadCompleteToEvenParity:`.
  * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:

>     * the VMI mutator webhook will copy the kubevirt CR annotation to the VMI annotations.

Author: why?

>     * the VMI mutator webhook will copy the kubevirt CR annotation to the VMI annotations.

Reviewer 2: Because it adds the cluster-wide solution.
The rest is almost the same as the per-VMI solution that is already implemented by https://github.com/kubevirt/kubevirt/pull/10593 and https://github.com/kubevirt/kubevirt/pull/10783.

>     * the VMI mutator webhook will copy the kubevirt CR annotation to the VMI annotations.

Author: The separation to phases is not really relevant anymore - since the changes to what is already merged is planned to be in one PR.

This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories
* As a cluster admin, I want to set the cluster-wide IsolatedEmulatorThread behavior for Kubevirt by setting a knob on HCO.

> * As a Kubevirt user, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

Reviewer 2: `As a Kubevirt user` -> `as a VM owner`.

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)

# Bug fix design proposal

> The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadEvenParity`:

Reviewer 2: Please consider `kubevirt.io/EmulatorThreadParityCompletion`.

Reviewer 2: Or `evenVCpusForIsolatedEmulatedThread` as suggested by @dankenigsberg.

Author: how about `kubevirt.io/SetEvenVCPUsForEmulatedThread`?

Reviewer 2: I personally prefer the first suggestion, but I think consulting a native speaker would help here.

Author: I asked @stu-gott, He said the either the suggested `kubevirt.io/EmulatorThreadEvenParity` or `kubevirt.io/EmulatorThreadCountEvenParity` might work. WDYT?

Author: chosen `kubevirt.io/EmulatorThreadCompleteToEvenParity`

```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  annotations:
    kubevirt.io/latest-observed-api-version: v1

>     kubevirt.io/storage-observed-api-version: v1

Reviewer 2: Please consider removing these annotations, as they are not related to the change.

    kubevirt.io/EmulatorThreadEvenParity:
spec:
  configuration:
    developerConfiguration:
      featureGates:
      - EmulatorThreadEvenParity
```

> This annotation will be supported only when an appropriate featureGate is enabled: EmulatorThreadEvenParity.

Author: DONE

When the following is configured:
1. kubevirt CR has the annotation.
2. kubevirt CR enabled the feature gate.
3. VM is configured with `isolateEmulatorThread: true`

then the virt-controller will:
1. Set the virt-launcher pod with an even amount of CPUs, meaning:

| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) | Notes                                                                       |
|-----------------------------------|------------------|-----------------------|-----------------------------------------------------------------------------|
| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of CPUs it requested | 

> 2. Copy the annotation to the VMI spec in order for it to be used on the virt-launcher, when setting the housekeeping cgroup. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.

Author: > In our discussions we have talked about making the VMI mutating webhook adding the annotation to the VMI (in order for the VMI-controller to just react to its presence). The VMI mutating webhook should add the annotation in case the following conditions are met:
> 
> 1. The FG is enabled.
> 2. The annotation is present on the KubeVirt CR.
> 3. The VMI is defined with IsolateEmulatorThread option enabled.

You're right, will fix. 
I've also set the conditions that needs to be met on every step.

The virt-launcher pod will then:
1. Try to assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin

The virt-handler will then:
1. Set the emulator Threads from the emulatorThreadPin to the housekeeping cgroup.

Pros:

> * Cleaner user experience. No prior knowledge of CPUs added behind the scene is needed.

Author: DONE

This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.

## User Stories
* As a cluster admin, I want to set the cluster-wide IsolatedEmulatorThread behavior for Kubevirt by setting a knob on HCO.

> * as a VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.

Reviewer 2: nit: `as a VM owner` -> `As a VM owner`

Author: DONE

## Repos
- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)

# Bug fix design proposal
The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadEvenParity`:
```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  annotations:
    kubevirt.io/EmulatorThreadEvenParity:
spec:
  configuration:
    developerConfiguration:
      featureGates:
      - EmulatorThreadEvenParity
```
This annotation will be supported only when the `EmulatorThreadEvenParity` featureGate is enabled.

## Virt-API:
When the following conditions are met:
1. kubevirt CR has the annotation.

> 2. kubevirt CR enabled the feature gate.

Reviewer 2: ```suggestion
2. The feature gate is enabled on the Kubevirt CR.
```

Please consider changing the order so the FG will be first, and the annotation on Kubevirt CR second.

Author: DONE

3. VM is configured with `isolateEmulatorThread: true`

then the VMI mutating-webhook will
1. Copy the annotation from the kubevirt CR to the VMI annotations. This is done in order for it to be used on the virt-launcher, when setting the housekeeping cgroup. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.

> 

Reviewer 2: There is an additional step, already implemented and merged on main:
The VMI creation admitter should allow the VMI to have the new annotation only if:
1. The FG is enabled.
2. The annotation on the Kubevirt CR is set.

Author: I think @vladikr was against it. am I wrong?

Reviewer 2: Please see: https://github.com/kubevirt/kubevirt/commit/bd8987871c2dc1b42dcece8a869c24e4e02c8d85

Author: I don't think it's the same..
In our case, the user (i.e. the VM owner) is not the one who's setting the annotation on the VMI.. it's an internal thing we do.

Reviewer 2: It is still possible for a user to manually set this annotation.
And IMO we should make sure that it is only used when the FG is enabled.

Author: talked offline, agreed the VMI create-admitter validation webhook should only allow the VMI if:
1. The feature gate is enabled on the Kubevirt CR
2. VMI is configured with `isolateEmulatorThread: true`

## Virt-controller:
When the following conditions are met:
1. kubevirt CR has the annotation.

> 2. kubevirt CR enabled the feature gate.

Reviewer 2: Since there is a check in the VMI creation addmitter, these are not necessary.

Author: I agree it's either this or that..

> 3. VM is configured with `isolateEmulatorThread: true`

Reviewer 2: VM -> VMI
And has the annotation.

Author: DONE

then virt-controller will:
1. Set the virt-launcher pod with an even amount of CPUs, meaning:

| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) | Notes                                                                       |
|-----------------------------------|------------------|-----------------------|-----------------------------------------------------------------------------|
| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of CPUs it requested |

## Virt-launcher:
When the following conditions are met:
1. VMI has the annotation.

> 2. VMI is configured with `isolateEmulatorThread: true`

Reviewer 2: Please consider reversing the order.
What leads the logic is the isolatedEmulatorThread option.

Author: true. DONE

# Implementation Phases
## Phase 1: Align CPU to SMT requirements when feature gate is enabled and annotation is set on the VMI.
* Virt-config:
  * Add support for the new feature gate: `EmulatorThreadEvenParity`.
* Virt-API:
  * Add support for the new alpha annotation: `kubevirt.io/EmulatorThreadEvenParity:`.
* Virt-controller:
  * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:
    * Add logic to Align the total number of CPUs to be even.

>     * Copy kubevirt CR annotation to VMI annotations.

Author: Fixing. DONE

## Phase 2: Adjust housekeeping cgroup to use the extra CPU added by the alignment, if such a one is available.
* Virt-launcher:

>   * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled

Author: DONE

# Implementation Phases
## Phase 1: Align CPU to SMT requirements when feature gate is enabled and annotation is set on the VMI.
* Virt-config:
  * Add support for the new feature gate: `EmulatorThreadEvenParity`.
* Virt-API:
  * Add support for the new alpha annotation: `kubevirt.io/EmulatorThreadEvenParity:`.
  * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:

>     * the VMI mutating-webhook will copy the kubevirt CR annotation to the VMI annotations.

Reviewer 2: There should also be a minor change to the VMI creation admitter to also consider the annotation on the Kubevirt CR.

Author: true. DONE

* Virt-controller:

>   * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:

Reviewer 2: Please see the comment above about performing this set in the VMI creation admitter.

Author: talked offline, DONE.
